### ISSUE TITLE
Issues in sync workflow

### ISSUE THREAD

**Issue by CheraHamza** [cecobask/imdb-trakt-sync#41]
Sync workflow kept failing after reaching the timeout. I set IMDB_TRACE to true, and it seems like some error is blocking the execution.
I keep getting those couple of log messages repeatedly until the timeout is reached and the sync workflow fails.

![Screenshot 2024-07-06 103502](https://github.com/cecobask/imdb-trakt-sync/assets/73847973/9e758501-9268-42d9-a1d3-a12b540a42e6)

**Comment by cecobask** [comment]
Hi @CheraHamza, your sync workflow is stuck at the stage after entering credentials and clicking the submit button.
The reason is more than likely that the IMDb website has prompted you to enter a captcha. This usually happens after a couple of failed attempts to authenticate with the provided credentials, or when you’ve logged in and out too frequently within a time period.
I recommend the following:
- Double check your IMDb credentials are correctly stored in the repository secrets
- Manually authenticate on the IMDb website which can help clear the captcha prompt
- Halt the sync workflow for a few hours before trying again

Let me know how it goes! I’m expecting some bugs might pop up here and there as I had to rewrite IMDb implementation.

**Comment by paviem** [comment]
Hi @cecobask, with IMDB_TRACE set to false, I get the following error (set to true, I get the same authorization error as @CheraHamza):
 
<img width="1038" alt="Screenshot 2024-07-06 at 3 13 15 PM" src="https://github.com/cecobask/imdb-trakt-sync/assets/128893079/d44d328a-d10d-4772-892a-9bbae9185a28">

**Comment by CheraHamza** [comment]
> * Manually authenticate on the IMDb website which can help clear the captcha prompt

Hi @cecobask, How can I manually authenticate though? 
If you mean the browser on my computer, then I am already signed in on IMDB

**Comment by aktur** [comment]
My manual workflow (set trace true) is stuck in the following:
```
{"time":"2024-07-06T14:31:51.22411+02:00","level":"INFO","source":{"function":"github.com/cecobask/imdb-trakt-sync/pkg/client.(*IMDbClient).RatingsExport","file":"/Users/michal/projects/imdb-trakt-sync/pkg/client/imdb.go","line":247},"msg":"exported ratings"}
[rod] 2024/07/06 14:31:51 [wait] load <page:CA7F4F6E>
[rod] 2024/07/06 14:31:51 [query] rod.element("div[data-testid='hero-list-subnav-export-button'] button")  <page:CA7F4F6E>
[rod] 2024/07/06 14:31:51 [query] rod.element("div[data-testid='hero-list-subnav-export-button'] button")  <page:CA7F4F6E>
[rod] 2024/07/06 14:31:52 [query] rod.element("div[data-testid='hero-list-subnav-export-button'] button")  <page:CA7F4F6E>
[rod] 2024/07/06 14:31:52 [query] rod.element("div[data-testid='hero-list-subnav-export-button'] button")  <page:CA7F4F6E>
[rod] 2024/07/06 14:31:54 [query] rod.element("div[data-testid='hero-list-subnav-export-button'] button")  <page:CA7F4F6E>
[rod] 2024/07/06 14:31:55 [query] rod.element("div[data-testid='hero-list-subnav-export-button'] button")  <page:CA7F4F6E>
[rod] 2024/07/06 14:31:56 [query] rod.element("div[data-testid='hero-list-subnav-export-button'] button")  <page:CA7F4F6E>
[rod] 2024/07/06 14:31:57 [query] rod.element("div[data-testid='hero-list-subnav-export-button'] button")  <page:CA7F4F6E>
[rod] 2024/07/06 14:31:58 [query] rod.element("div[data-testid='hero-list-subnav-export-button'] button")  <page:CA7F4F6E>
[rod] 2024/07/06 14:31:59 [query] rod.element("div[data-testid='hero-list-subnav-export-button'] button")  <page:CA7F4F6E>
[rod] 2024/07/06 14:32:00 [query] rod.element("div[data-testid='hero-list-subnav-export-button'] button")  <page:CA7F4F6E>
[rod] 2024/07/06 14:32:01 [query] rod.element("div[data-testid='hero-list-subnav-export-button'] button")  <page:CA7F4F6E>
[rod] 2024/07/06 14:32:02 [query] rod.element("div[data-testid='hero-list-subnav-export-button'] button")  <page:CA7F4F6E>
[rod] 2024/07/06 14:32:03 [query] rod.element("div[data-testid='hero-list-subnav-export-button'] button")  <page:CA7F4F6E>
[rod] 2024/07/06 14:32:04 [query] rod.element("div[data-testid='hero-list-subnav-export-button'] button")  <page:CA7F4F6E>
[rod] 2024/07/06 14:32:05 [query] rod.element("div[data-testid='hero-list-subnav-export-button'] button")  <page:CA7F4F6E>
[rod] 2024/07/06 14:32:06 [query] rod.element("div[data-testid='hero-list-subnav-export-button'] button")  <page:CA7F4F6E>
[rod] 2024/07/06 14:32:07 [query] rod.element("div[data-testid='hero-list-subnav-export-button'] button")  <page:CA7F4F6E>
[rod] 2024/07/06 14:32:08 [query] rod.element("div[data-testid='hero-list-subnav-export-button'] button")  <page:CA7F4F6E>
...
```
until timeout...

**Comment by cecobask** [comment]
When the `IMDB_TRACE` configuration value is set to `true`, the syncer outputs extra information about the activities happening in the headless browser. @paviem, you're experiencing the same issue in both cases, but when tracing is disabled you don't get the detailed logs output.

The syncer was initially set up to wait on two conditions after the submit button on the authentication page is pressed:
- successful authentication
- failed authentication due to invalid credentials

I did not consider that captcha prompts would show up after a couple of failed authentication attempts. To address this issue, in the latest commit I added a safeguard which would report if a captcha prompt had hindered the authentication flow and cancel the sync early. Without the latest changes, the sync can get stuck waiting for one of the above two conditions until the timeout kicks in.

I also added bot detection prevention, but that unfortunately will not eliminate the possibility of captcha prompts appearing.

In summary, if anybody is experiencing this issue they can do the following:
- sync their fork with the latest changes
- ensure the IMDb credentials stored as repository secrets are valid (**this is very important**)
- run the syncer locally with `IMDB_HEADLESS` config value set to `false`
